designs/comprehensions-design.md

```diff
@@ -0,0 +1,216 @@
+
+## Week 7:
+### Test cases implemented and passed:
+All test cases were written in the tests/comprehension.test.ts file. Upon running this file, 15/15 teste cases passed. The following are the test cases that pass and produce the required output:
+
+1. simple comprehension output with min and max range
+```j: int = 2```
+```print([j for a in range(5,7)])```
+Output: ```['2', '2']```
+
+2. simple comprehension output with only max range
+```j: int = 7```
+```print([j for b in range(5)])```
+Output: ```['7', '7', '7', '7', '7']```
+
+3. simple comprehension output with bool values
+```j: bool = True```
+```print([j for c in range(1,5)])```
+Output: ```['True', 'True', 'True', 'True']```
+
+4. simple comprehension output with expr values
+```j: int = 5```
+```print([j*2 for d in range(1,5)])```
+Output: ```['10', '10', '10', '10']```
+
+5. simple comprehension output using iterable class methods
+```print([e for e in range(1,5)])```
+Output: ```['1', '2', '3', '4']```
+
+6. simple comprehension output using iterable class methods and expr values
+```print([f*3 for f in range(1,5)])```
+Output: ```['3', '6', '9', '12']```
+
+7. simple comprehension output with if condition
+```print([g for g in range(1,5)])```
+Output: ```['1', '2', '4']```
+
+8. simple comprehension output with bool binop expr values and if condition
+```j: int = 3```
+```print([j<i for i in range(6) if i>3])```
+Output: ```['False', 'False', 'True']```
+
+9. simple comprehension output with function call as expr values
+```def f(x:int)->int: return x*5```
+```j: int = 5```
+```print([f(j) for l in range(5)])```
+Output: ```['25', '25', '25', '25', '25]```
+
+10. simple comprehension output with step
+```print([k for k in range(10,20,2)])```
+Output: ```['10', '12', '14', '16', '18']```
+
+11. invalid expression in comprehension
+```print([j for a in range(1,5) if a!=1])```
+Output: ```Error: j is undefined```
+
+12. invalid range in comprehension
+```j: int = 2```
+```print([j for a in range(2,1)])```
+Output: ```Error```
+
+13. only if condition allowed in comprehension
+```j: int = 2```
+```print([j for a in range(1,5) for a!=1])```
+Output: ```Error```
+
+14. invalid condition in comprehension
+```j: int = 2```
+```print([j for a in range(1,5) if a+2])```
+Output: ```Error```
+
+15. invalid iterable in comprehension
+```j: int = 2```
+```k: int = 3```
+```print([j for a in k if a!=2])```
+Output: ```Error```
+
+Test cases that does not work yet is the following:
+
+1. ```print([[j for j in range(3)] for i in range(3)])```
+2. 
+```
+ a : List = None
+a = [i for i in range(5,10)]
+print(a)
+```
+In case of the second test case, we skipped it because as suggested by proffesor, we plan on using the implementation by the lists team. For this week, we only worked on iterables using the range class.
+
+## Limitation so far (plan on fixing in week 8)
+
+1. We made an assumption that the list comprehension expressions will only be used globally, and not locally inside any function.
```

Comment:
What changes do you need to make to generalize the context of comprehensions?

Reply:
We have fixed this. In the previous version of our code, we were creating range objects in the type check file that were global in nature. We have a more generalized code in type check now, that we are flattening in lower.ts. Th limitation is no longer valid and has been updated.